### PR TITLE
Add validation when update device profile

### PR DESCRIPTION
**Pull request by AnthonyMBonafide**
Fix issue #1494

Add checks before updating device profile and replace direct calls to the DBClient with the operator pattern to allow for easier testing.

Update the restUpdateDeviceProfile rest handler to check if there are
associated devices or provision watchers before performing an update. If
there are associations for the device profile a Conflict(409) is
returned to the user and the update is not persisted. Update the successful response to return a 204(No Content) with no body rather than a "true"

Add new `UpdateDeviceProfileExecutor` which provides functionality for
updating device profiles by using the operator pattern. Also, move all
business logic to the `UpdateDeviceProfileExecutor` for segregation of
responsibilities. The REST handler is responsible for HTTP/REST logic
like serializing the request body, where the
`UpdateDeviceProfileExecutor` is responsible for interacting with the
database and ensuring the device profile which is being updated is in
the correct state before performing the update. For example, no devices
nor provision watchers are associated with the device profile.

Update the `restUpdateDeviceProfile` rest handler to use the new
operator.

Add new error types for the different errors the
`UpdateDeviceProfileExecutor` may encounter during processing. This will
allow for type checking for more insight on the underlying error

Update the `restUpdateDeviceProfile` to handle the different types of
errors the `UpdateDeviceProfileExecutor` may return

Add new `ProfileIdExecutor` for obtaining devices by a profile ID using
the operator pattern.

Update `notifyProfileAssociates` function to use the new operator

Added necessary mocks for new interfaces

Remove dead code which checked for duplicate command names within
DeviceProfile. This is dead code due to the validation logic already
handling this scenario and will result in the marshaling process to
short-circuit and result in a 400(Bad Request) response to the user

Add tests for `restUpdateDeviceProfile` REST handler, `UpdateDeviceProfileExecutor`, and `UpdateDeviceProfileExecutor`

Signed-off-by: Anthony M. Bonafide <AnthonyMBonafide@gmail.com>

### CODE REVIEW

**Comment by AnthonyMBonafide** [comment]
TLDR:
Added the necessary checks as per the associated issue. and moved all the business logic into an operator. Also added tests.

File: internal/core/metadata/rest_deviceprofile.go [hunk 1]
```diff
@@ -97,119 +101,33 @@ func restUpdateDeviceProfile(w http.ResponseWriter, r *http.Request) {
...
-		to.CoreCommands = from.CoreCommands
-	}
-
-	return nil
-}
-
-// Check for duplicate names in device profiles
-func checkDuplicateProfileNames(dp models.DeviceProfile, w http.ResponseWriter) error {
-	profiles, err := dbClient.GetAllDeviceProfiles()
+	err = notifyProfileAssociates(dp, dbClient, http.MethodPut)
```
Comment (AnthonyMBonafide) [inline internal/core/metadata/rest_deviceprofile.go]: I did not include tests to cover the `notifyProfileAssociates` function. This function is tightly coupled to the `http` client. I did not want to complicate this PR anymore by refactoring further to decouple the HTTP client, so I left those tests out.

Comment (tsconn23) [reply internal/core/metadata/rest_deviceprofile.go]: Yep, that's fine. When I created the original test for notifications in the AddNewDevice path, I knew there would be other areas that would need a similar abstraction. Could be they'll all converge at some point.

File: internal/core/metadata/errors/types.go [hunk 2]
```diff
@@ -106,3 +106,47 @@ func (e ErrItemNotFound) Error() string {
...
+	id   string
+}
+
+func (e ErrDeviceProfileNotFound) Error() string {
+	if e.id != "" {
+		return fmt.Sprintf("device profile with id '%s' not found", e.id)
+	} else if e.name != "" {
+		return fmt.Sprintf("device profile with name '%s' not found", e.name)
+	}
+	return "device profile not found"
```
Comment (tsconn23) [inline internal/core/metadata/errors/types.go]: Suggest reduce the body of this function to just be
```
return fmt.Sprintf("device profile not found -- id: '%s' name: '%s'", e.id, e.name)
```

File: internal/core/metadata/errors/types.go [hunk 3]
```diff
@@ -106,3 +106,47 @@ func (e ErrItemNotFound) Error() string {
...
+	description string
+}
+
+func (e ErrDeviceProfileInvalidState) Error() string {
+	if e.id != "" {
+		return fmt.Sprintf("device profile with id '%s' is in an invalid state. %s", e.id, e.description)
+	} else if e.name != "" {
+		return fmt.Sprintf("device profile with name '%s' is in an invalid state. %s", e.name, e.description)
+	}
+	return "device profile invalid state"
```
Comment (tsconn23) [inline internal/core/metadata/errors/types.go]: Same as above, can be collapsed to one line.

File: internal/core/metadata/operators/device/get.go [hunk 4]
```diff
@@ -15,10 +15,11 @@
 package device
 
 import (
-	"github.com/edgexfoundry/edgex-go/internal/core/metadata/errors"
-	"github.com/edgexfoundry/edgex-go/internal/pkg/config"
 	"github.com/edgexfoundry/go-mod-core-contracts/clients/logger"
 	contract "github.com/edgexfoundry/go-mod-core-contracts/models"
+
+	"github.com/edgexfoundry/edgex-go/internal/core/metadata/errors"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/config"
```
Comment (tsconn23) [inline internal/core/metadata/operators/device/get.go]: Not sure why these were moved as `e` comes before `g`

File: internal/core/metadata/operators/device/get_test.go [hunk 5]
```diff
@@ -15,25 +15,29 @@
 package device
 
 import (
-	"github.com/pkg/errors"
 	"testing"
 
-	"github.com/edgexfoundry/edgex-go/internal/core/metadata/operators/device/mocks"
-	"github.com/edgexfoundry/edgex-go/internal/pkg/config"
+	"github.com/pkg/errors"
```
Comment (tsconn23) [inline internal/core/metadata/operators/device/get_test.go]: This should be last in the ordering, alphabetizing everything after github.com. Also edgex-go comes before go-mod-core-contracts.

File: internal/core/metadata/operators/device/profile/update.go [hunk 6]
```diff
@@ -0,0 +1,141 @@
+package profile
+
+import (
+	"strconv"
+
+	contract "github.com/edgexfoundry/go-mod-core-contracts/models"
+
+	errors2 "github.com/edgexfoundry/edgex-go/internal/core/metadata/errors"
```
Comment (tsconn23) [inline internal/core/metadata/operators/device/profile/update.go]: Shouldn't need the `errors2` alias here.

File: internal/core/metadata/operators/device/profile/update.go [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+	}
+
+	d, err := n.database.GetDevicesByProfileId(to.Id)
+	if err != nil {
+		return contract.DeviceProfile{}, err
+	}
+
+	if len(d) > 0 {
+		// LoggingClient.Debug(strconv.Itoa(len(d))+"devices are associated with the device profile named: " + to.Name)
+		// http.Error(w, "Conflict", http.StatusConflict)
```
Comment (tsconn23) [inline internal/core/metadata/operators/device/profile/update.go]: comments can be removed

File: internal/core/metadata/operators/device/profile/update.go [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+func NewUpdateDeviceProfileExecutor(db DeviceProfileUpdater, dp contract.DeviceProfile) UpdateDeviceProfileExecutor {
+	return updateDeviceProfile{
+		database: db,
+		dp:       dp}
+}
+
+// Update the fields of the device profile
+// to - the device profile that was already in Mongo (whose fields we're updating)
+// from - the device profile that was passed in with the request
+func (n updateDeviceProfile) updateDeviceProfileFields(from contract.DeviceProfile, to *contract.DeviceProfile) error {
```
Comment (tsconn23) [inline internal/core/metadata/operators/device/profile/update.go]: I know this is a previously defined function and you didn't write it, just moved it into the operator from the router. I _**cannot stand**_ functions that mutate their parameters. I cannot tell you how many bugs I've had to root cause because of code written this way. Suggest this be refactored to return the new DeviceProfile rather than mutate the `*to` parameter. The caller should be changed to expect the new return as well.

File: internal/core/metadata/operators/device/profile/update.go [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+	if from.Origin != 0 {
+		to.Origin = from.Origin
+	}
+
+	if from.Name != "" {
+		to.Name = from.Name
+		// Names must be unique for each device profile
+		if err := n.checkDuplicateProfileNames(*to); err != nil {
+			return err
+		}
```
Comment (tsconn23) [inline internal/core/metadata/operators/device/profile/update.go]: In my opinion, this check should be moved into Execute() and performed before this function is even called. Once we get to invoking this function, all checks should be complete and the function does nothing but assign values to the new instance.

_Something to think about for later -- the creation of a new instance from old could actually be a function of the DeviceProfile model itself..._

File: internal/core/metadata/rest_deviceprofile.go [hunk 10]
```diff
@@ -20,10 +20,14 @@ import (
...
 
-	"github.com/edgexfoundry/edgex-go/internal/pkg/db"
 	"github.com/edgexfoundry/go-mod-core-contracts/models"
 	"github.com/gorilla/mux"
 	"gopkg.in/yaml.v2"
+
+	errors2 "github.com/edgexfoundry/edgex-go/internal/core/metadata/errors"
+	"github.com/edgexfoundry/edgex-go/internal/core/metadata/operators/device"
+	"github.com/edgexfoundry/edgex-go/internal/core/metadata/operators/device/profile"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/db"
```
Comment (tsconn23) [inline internal/core/metadata/rest_deviceprofile.go]: Sorting's a little weird here. See other comments as well. All github.com imports should be alphabetized. Ignore aliases.

**Comment by AnthonyMBonafide** [comment]
> One thing I'd ask is to not create a `device/profile` directory.

Should I put everything in the `device` directory?

File: internal/core/metadata/operators/device/profile/update.go (hunk 8 above)
Comment (AnthonyMBonafide) [reply internal/core/metadata/operators/device/profile/update.go]: :+1: Excellent point and I agree

**Comment by tsconn23** [comment]
> > One thing I'd ask is to not create a `device/profile` directory.
> 
> Should I put everything in the `device` directory?

No, create a `metadata/operators/device_profile` directory

**Comment by AnthonyMBonafide** [comment]
Recheck

**Comment by tsconn23** [comment]
@AnthonyMBonafide One last thing on this -- Can you add the new 409 status being returned for this route to the RAML please? Then it should be good to go.

https://github.com/edgexfoundry/edgex-go/blob/42b502f5bb4ec663984d8826656a4b84953494ce/api/raml/core-metadata.raml#L803
